### PR TITLE
Favorites

### PR DESCRIPTION
**Pull request by bbelderbos**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +91,87 @@ Copied 🐍 💪 🎉 💓 🔥 to clipboard
 
 After sourcing your .zshrc you can now get emojis copied to your clipboard fast using `emo bicep`, `emo tada` etc.
 
+### Preferred emojis
```
Comment (bbelderbos) [inline README.md:101]: Started here!

Nice for a new article: _Documentation-driven development in practice_

File: tests/test_script.py [hunk 2]
```diff
@@ -1,13 +1,42 @@
...
+
+@pytest.fixture(scope="session")
+def add_preferences():
+    prefs_file = os.environ.get("EMOJI_PREFERENCES")
+    assert prefs_file is not None
+
+    prefs_file = Path(prefs_file)
+    prefs_file.write_text(PREFERENCE_FILE_CONTENT)
+
+    yield
```
Comment (bbelderbos) [inline tests/test_script.py:36]: love this with fixtures: setup - yield - teardown

File: requirements.in [hunk 3]
```diff
@@ -0,0 +1,8 @@
+pyperclip
```
Comment (bbelderbos) [inline requirements.in:1]: using pip-tools for every project now

File: pytest.ini [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[pytest]
+env =
+    EMOJI_PREFERENCES=.preferences-test
```
Comment (bbelderbos) [inline pytest.ini:3]: TIL :)

File: emojisearcher/script.py [hunk 5]
```diff
@@ -21,39 +23,60 @@ def clean_non_emoji_characters(emoji: str) -> str:
...
+    if preferences is None:
+        preferences = load_preferences()
+
     matches = []
+    is_preference_emoji = False
     for word in words:
-        emojis = get_emojis_for_word(word.rstrip(SIGNAL_CHAR))
-        if len(emojis) == 0:
-            continue
+        if word in preferences:
```
Comment (bbelderbos) [inline emojisearcher/script.py:49]: testing showed me I had to move this up (see commit history in PR)

File: emojisearcher/script.py [hunk 6]
```diff
@@ -21,39 +23,60 @@ def clean_non_emoji_characters(emoji: str) -> str:
 
 
 def get_matching_emojis(
-    words: list[str], interactive: bool = False
+    words: list[str],
+    *,
+    preferences: dict[str, str] | None = None,
```
Comment (bbelderbos) [inline emojisearcher/script.py:28]: love the new typing syntax (3.10)

File: README.md [hunk 7]
```diff
@@ -1,10 +1,14 @@
...
+## Pybites Emoji Searcher
 
-Automate the boring stuff: I have been googling emojis and manually copying them to my clipboard (it's especially painful when posting to Facebook).
+I have been googling emojis and manually copying them to my clipboard.
 
-So it was time to write a script to look up emojis by text from the command line and copy them to the clipboard.
+Except for Slack + GitHub, there the `:` + autocomplete works great. For other tools, for example Facebook or plain blog / email writing, I needed a better way.
 
-By default it takes the first match in case there are multiple matching emojis. However if you append a dot (.) to a word you get to choose which emoji gets copied.
+So here is a tool to look up emojis by text from the command line and automatically copy matching ones to the clipboard (using the awesome [pyperclip](https://pyperclip.readthedocs.io/en/latest/) tool).
```
Comment (spackle0) [inline README.md:7]: 👍

File: README.md [hunk 8]
```diff
@@ -1,10 +1,14 @@
...
+
+By default it takes the first match in case there are multiple matching emojis. However if you append a dot (.) to a word you get to choose which emoji gets copied. You can also use a `.preferences` file to store overriding emojis or ones this tool does not provide.
+
+I hope you enjoy this tool and don't hesitate to reach out to me by email: bob@pybit.es or just open an issues / open a PR if you see any opportunity for improvements.
```
Comment (spackle0) [inline README.md]: "open an issue"
(singular)

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+lint:
+	flake8 emojisearcher tests
+
+.PHONY: typing
+typing:
+	mypy emojisearcher tests
+
+.PHONY: cov
+cov:
+	pytest --cov=emojisearcher --cov-report=term-missing
```
Comment (spackle0) [inline Makefile]: `--cov-fail-under=80` for example.

File: README.md [hunk 10]
```diff
@@ -67,13 +74,10 @@ End a word with a . if you want to select an emoji if there are multiple
...
 > q
 Bye
-
-
-# running the tests
-(venv) $ pip install -r requirements-dev.txt
-(venv) $ pytest
 ```
 
+### Easy of use: make a shell alias
```
Comment (spackle0) [inline README.md]: "Ease of use"

File: README.md [hunk 11]
```diff
@@ -94,10 +98,111 @@ Copied 🐍 💪 🎉 💓 🔥 to clipboard
...
+
+# run flake8 and mypy
+(venv) $ make lint
+(venv) $ make typing
+```
+
 ### Rich
 
-Update: I am using `rich` now to retrieve a list of emojis, it seems a bit more accurate (e.g. our beloved 🎉 emoji). I will be enhancing the cli interface with this awesome library ...
+Originally Around 0.0.5 we started `rich` now to retrieve a list of emojis, it seems a bit more accurate (e.g. our beloved tada 🎉 emoji was missing!)
```
Comment (spackle0) [inline README.md]: "we started using `rich` to retrieve"

File: emojisearcher/script.py [hunk 12]
```diff
@@ -21,39 +23,60 @@ def clean_non_emoji_characters(emoji: str) -> str:
...
 
 def get_emojis_for_word(
     word: str, emoji_mapping: dict[str, str] = EMOJI_MAPPING
 ) -> list[str]:
-    # TODO: mypy says "Incompatible types in assignment"
     return [emo for name, emo in emoji_mapping.items() if word in name]
 
 
-def user_select_emoji(emojis: list[str]) -> str:
+def user_select_emoji(emojis: list[str]) -> str | None:
```
Comment (spackle0) [inline emojisearcher/script.py:79]: 👍

File: emojisearcher/script.py [hunk 13]
```diff
@@ -67,6 +90,9 @@ def user_select_emoji(emojis: list[str]) -> str:
         except IndexError:
             print(f"{user_input} is not a valid option.")
             continue
+        except KeyboardInterrupt:
```
Comment (spackle0) [inline emojisearcher/script.py:93]: Nice and clean!

File: pytest.ini (hunk 4 above)
Comment (spackle0) [reply pytest.ini:3]: Same!

File: requirements.in (hunk 3 above)
Comment (spackle0) [reply requirements.in:1]: I'd love to know why.

Comment (bbelderbos) [reply requirements.in:1]: Easy to use and the requirements.txt it generates has more info how dependencies relate to each other (similar to pipdeptree).

File: Makefile (hunk 9 above)
Comment (bbelderbos) [reply Makefile]: TIL, thanks!

File: README.md (hunk 10 above)
Comment (bbelderbos) [reply README.md]: Thanks for the spelling fixes, committing those now

File: emojisearcher/script.py (hunk 13 above)
Comment (bbelderbos) [reply emojisearcher/script.py:93]: Thanks

File: requirements.in (hunk 3 above)
Comment (spackle0) [reply requirements.in:1]: I'll see how that compares to poetry. Thanks.

**Pull request merged**